include/vcpkg/binarycaching.h

```diff
@@ -9,53 +9,131 @@
 
 #include <vcpkg/packagespec.h>
 
+#include <iterator>
+#include <string>
 #include <unordered_map>
+#include <vector>
 
 namespace vcpkg
 {
-    struct MergeBinaryProviders;
-
     enum class RestoreResult
     {
-        missing,
-        success,
-        build_failed,
+        unavailable,
+        restored,
+    };
+
+    enum class CacheAvailability
+    {
+        unavailable,
+        available,
+    };
+
+    enum class CacheStatusState
+    {
+        unknown,   // the cache status of the indicated package ABI is unknown
+        available, // the cache is known to contain the package ABI, but it has not been restored
+        restored,  // the cache contains the ABI and it has been restored to the packages tree
+    };
+
+    struct IBinaryProvider;
+
+    struct CacheStatus
+    {
+        CacheStatus() noexcept;
+        CacheStatus(const CacheStatus&);
+        CacheStatus(CacheStatus&&) noexcept;
+        CacheStatus& operator=(const CacheStatus&);
+        CacheStatus& operator=(CacheStatus&&) noexcept;
+        ~CacheStatus();
+
+        bool should_attempt_precheck(const IBinaryProvider* sender) const noexcept;
```

Comment:
As a general note, these apis feel a bit unfortunate to me (passing `this` to most functions). However, I'm not sure of a cleaner alternative considering the desire to reuse the cache between plans.

The only thing that comes to mind is the full-fortran style of:
```cpp
void IBinaryProvider::prefetch(
    View<InstallAction>,
    View<int> cache_offsets,
    View<enum SharedCacheState>,
    View<enum PrivateCacheState>)
```
Where the owning `BinaryCache` then has:
```cpp
std::unordered_map<PackageSpec, int> offset_map;
std::vector<enum SharedCacheState> shared_state;
std::vector<std::vector<enum PrivateCacheState>> private_states;
std::vector<IBinaryProvider*> providers;
```
with further
```cpp
enum class SharedCacheState : char {
    all_unavailable, // start here, this indicates that every provider so far has confirmed they do not have the package
    any_unknown, // transition to this from all_unavailable if a provider is unable to precheck
    restored, // transition to this if a provider claimed the package
};
enum class PrivateCacheState : char {
    unknown, // I have not prechecked for this package
    unavailable, // I have prechecked for this package and I don't have it
    available, // I have prechecked for this package and I do have it
};
```

Though I don't think this suggestion is sufficiently better to warrant the rewrite.

Reply:
I'm not even really convinced that's better -- I think the 'this' use and dual purpose of both uniquely identifying the matching provider(s) and making finding the provider(s) themselves cheap and easy are good properties.